ports/stm32/modmachine.c

```diff
@@ -351,7 +351,14 @@ MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(machine_freq_obj, 0, 4, machine_freq);
 // This executies a wfi machine instruction which reduces power consumption
 // of the MCU until an interrupt occurs, at which point execution continues.
 STATIC mp_obj_t machine_idle(void) {
+    #if defined(STM32L1)
+    mp_uint_t irq_state;
+    irq_state = MICROPY_BEGIN_ATOMIC_SECTION();
```

Comment:
why did you need to take the atomic state here?

Reply:
For STM32L1, IRQ state may be changed after __WFI() only in MCU debug component (DBGMCU) is enabled.
This causes unexpected HardFault. To prevent it, I needed to save and restore IRQ state beetween __WFI().

Usually DBGMCU is enabled only when debugging with OpenOCD or using STM32CubeProgrammer with enableing 'Debug in Low Power mode'.

Reply:
I'll have to understand more about this.  Is it only for the STM32L1xx series?  Is there an application note (or some documentation) about why and when it's needed?

I don't want to add this for all ports, especially in the `MICROPY_EVENT_POLL_HOOK` macro.

Reply:
> Is it only for the STM32L1xx series?

Maybe, yes. I didn't meet such HardFault with NUCLEO-F446RE.
So I added these implementation for STM32L1 only.

> Is there an application note (or some documentation) about why and when it's needed?

I'm sorry that I couldn't find any documentation, but there are some similar issues.
For STM32L1xx, program counter seems to corrupt after wfi when irq and DBGMCU are enabled.
https://community.st.com/s/question/0D50X00009XkbFuSAJ/wfi-in-run-mode-causing-hard-fault
https://github.com/probe-rs/probe-rs/issues/350
https://github.com/RIOT-OS/RIOT/issues/11820
https://github.com/RIOT-OS/RIOT/pull/11830

> I don't want to add this for all ports, especially in the MICROPY_EVENT_POLL_HOOK macro.

This issue is due to enabling DBGMCU, so disabling DBGMCU instead of taking atomic state can also prevent this issue.
To disable it, execute
`DBGMCU->CR &= ~(DBGMCU_CR_DBG_SLEEP | DBGMCU_CR_DBG_STOP | DBGMCU_CR_DBG_STANDBY);`
during initialization of STM32L1.
Even if disabling DBGMCU, I could debug by using ST-LINK or OpenOCD.
Should I fix it?

Reply:
To be able to move forward with this PR (merge it) I suggest disabling DBGMCU during initialisation of the STM32L1xx, and not changing the WFI behaviour of other ports.

That will at least keep things the same for existing MCUs, and get it working on STM32L152.

Then we can possible fix/improve things in the future.

Reply:
Thank you for reply. I added some commits.

- I removed taking atomic state between __WFI() and add disabling DBGMCU at initialization (49c66271).
- When using STM32CubeProgrammer to upload firmware, the user must disable 'Debug in Low Power Mode' (dLPM).
Also, STM32CubeProgrammer often cannot detect NUCLEO-L152RE, so adding mode 'Under Reset' (mode=UR).
To explain it, I added deploy.md for NUCLEO-L152RE (8d520538, d248e3b7).
- While writing deploy.md, I also noticed that machine.bootloader() was not working.
Now I fixed it, so please include this commit (e14f2627) to this PR.
Sorry for inconvenience.